src/base/static/components/translator.js

```diff
@@ -0,0 +1,73 @@
+import React, { useState, useEffect } from "react";
+import PropTypes from "prop-types";
+import qs from "qs";
+
+import { LanguageConsumer } from "../context/language";
+
+const Translator = props => {
+  const [translatedMsg, setTranslatedMsg] = useState(props.msg);
```

Comment:
I'm not sure if it makes sense to use state here to store translated content. Aren't we duplicating what's in the ducks? Should we instead be updating the content in the ducks themselves?

Reply:
Yeah, I'm not sure either. But I think it makes sense to keep the default language as a "source of truth".

It will also allow us to cache multiple translations in this component by updating our state to be a mapping of language codes to translation strings.

Reply:
I'm realizing now that one drawback to this approach is that any cached translations in the `T` component will be lost when instances of that component unmount.

A common example of this in practice would be a user switching back and forth between templates (like the list and the map). One each template switch all static strings would be re-translated via API calls.

Reply:
It seems like there are 2 options:

1. Keep storing the translations in component state, and if it's a cache miss, we re-fetch the translation.

2. Or, we can store the translations in the context. If so, we should structure our context similar to React i18next, so that it's easy to look it up by component. For example:

```js
{
  currentLanguage: this.props.currentLanguage,
  defaultLanguage: this.props.defaultLanguage,
  es: {
    InputFormCategoryButton: {
      "label": "añadir un lugar",
    },
    PlaceForm: {
      "title": "un titulo",
      "description": "la descripción",
    }
  },
}
```

But that will require the component to pass in the keys, like `InputFormCategoryButton` and `label`. Or we can figure out a way to automate passing in the keys.

Reply:
I think the cache misses will be substantial enough that it's worth exploring another solution. I was thinking along the lines of what you outline above, so I'll try that and see how it works.

I might skip automatically generating keys for now though. We'd also need to handle cases where numerous instances of the same component get rendered with different translatable strings. So maybe we could manually pass in the component namespace and unique key:

```
arrayOfStuff.map(item => {
  return (
    <SomeComponent
        key={item.property}
        gTranslateNamespace="SomeComponent"
        gTranslateKey={item.property} />
  )
}
```

where `item.property` is some unique identifier also suitable for use as a React `key`.

Reply:
Though we're also pretty close to just replicating i18next's resource store at this point, so maybe we should just update the resource store with dynamically translated content as needed...